fuel-core/src/database/utils.rs

```diff
@@ -0,0 +1,218 @@
+    ) -> Self {
+        Self {
+            owner,
+            asset,
+            exclude,
+            database,
+        }
+    }
+
+    /// Returns the iterator over all valid(spendable, allowed by `exclude`) resources of the `owner`
+    /// for the `asset_id`.
+    // TODO: Optimize this by creating an index
+    pub fn unspent_resources(
+        &self,
+    ) -> impl Iterator<Item = Result<Resource<Cow<Coin>, Cow<Message>>, Error>> + '_ {
+        let coins_iter = self
+            .database
+            .owned_coins_utxos(self.owner, None, None)
+            .filter_ok(|id| {
+                if let Some(exclude) = self.exclude {
+                    !exclude.utxos.contains(id)
+                } else {
+                    true
+                }
+            })
+            .map(|res| {
+                res.map(|id| {
+                    let coin = Storage::<UtxoId, Coin>::get(self.database, &id)?
+                        .ok_or(KvStoreError::NotFound)?;
+
+                    Ok::<_, KvStoreError>(Resource::Coin { id, fields: coin })
+                })
+            })
+            .map(|results| Ok(results??))
+            .filter_ok(|coin| {
+                if let Resource::Coin { fields, .. } = coin {
+                    fields.asset_id == self.asset.id
+                        && fields.status == CoinStatus::Unspent
+                } else {
+                    true
+                }
+            });
+
+        let messages_iter = self
+            .database
+            .owned_message_ids(self.owner, None, None)
+            .filter_ok(|id| {
+                if let Some(exclude) = self.exclude {
+                    !exclude.messages.contains(id)
+                } else {
+                    true
+                }
+            })
+            .map(|res| {
+                res.map(|id| {
+                    let message = Storage::<MessageId, Message>::get(self.database, &id)?
+                        .ok_or(KvStoreError::NotFound)?;
+
+                    Ok::<_, KvStoreError>(Resource::Message {
+                        id,
+                        fields: message,
+                    })
+                })
+            })
+            .map(|results| Ok(results??))
+            .filter_ok(|message| {
+                if let Resource::Message { fields, .. } = message {
+                    fields.fuel_block_spend.is_none()
+                } else {
+                    true
+                }
+            });
+
+        coins_iter
+            .chain(messages_iter.take_while(|_| self.asset.id == ChainConfig::BASE_ASSET))
+    }
+}
+
+/// The id of the resource.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ResourceId {
+    Utxo(UtxoId),
+    Message(MessageId),
+}
+
+/// The primary type of spent or not spent resources(coins, messages, etc). The not spent resource
+/// can be used as a source of information for the creation of the transaction's input.
+#[derive(Debug)]
+pub enum Resource<C, M> {
```

Comment:
Since the schema has its own analog to resource, why include these generics here?

Reply:
To avoid unnecessary `Copy` and `Clone`. The query to DB returns `Resource<Cow<Coin>, Cow<Message>>`. But the graphql query returns `Resource<Coin, Message>`.

Reply:
> Since the schema has its own analog to resource

It looks like the schema doesn't rely on this enum, since it has it's own: https://github.com/FuelLabs/fuel-core/blob/feature/coins-to-spent-messages/fuel-core/src/schema/resource.rs#L52

Couldn't we just simplify this type by making it always a Cow?

Reply:
No, because `Cow` requires lifetime from DB, and in the `resources_to_spend` we can't return `Vec<Vec<Resource<Cow, Cow>>>`=(

Reply:
Do we really need `Cow`? Or is it always owned once deserialized from the DB anyways? Even though the storage trait uses Cow, I don't think it would ever be borrowed unless we had some kind of zero-copy deserialization going on.

Reply:
Yes, if we use RocksDB, in this case, it is useless=) I tried to be consistent with our API. `Database` returns `Cow`, and in some cases (maybe in the future), it can be useful. 

If you think it adds unjustified complexity, I can remove it=)

Reply:
eh it's not a big deal. I just prefer to find ways to hide generics when possible since it can get out of hand sometimes (like in substrate codebases lol)